### PR TITLE
[Batched Generators] This PR adds generators that are useful to make batched generation fully reproducible 

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR adds two nice features:
- Test that verifies that batched output and single output is identical
- Adds a nice doc page explaining how "batched" generators can be used for nice prompt engineering

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Tests are flaky I think.

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py [hunk 1]
```diff
@@ -310,8 +310,8 @@ def __call__(
                 1`. Higher guidance scale encourages to generate images that are closely linked to the text `prompt`,
                 usually at the expense of lower image quality.
             generator (`torch.Generator`, *optional*):
-                A [torch generator](https://pytorch.org/docs/stable/generated/torch.Generator.html) to make generation
-                deterministic.
+                One or a list of [torch generator(s)](https://pytorch.org/docs/stable/generated/torch.Generator.html)
+                to make generation deterministic.
             latents (`jnp.array`, *optional*):
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py:315]: misplaced generator docstring in a flax pipeline

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py [hunk 2]
```diff
@@ -359,16 +373,24 @@ def prepare_latents(self, image, timestep, batch_size, num_images_per_prompt, dt
...
-        # add noise to latents using the timesteps
-        noise = torch.randn(init_latents.shape, generator=generator, device=device, dtype=dtype)
+            init_latents = torch.cat([init_latents], dim=0)
+
+        rand_device = "cpu" if device.type == "mps" else device
+        shape = init_latents.shape
+        if isinstance(generator, list):
+            shape = (1,) + shape[1:]
+            noise = [
+                torch.randn(shape, generator=generator[i], device=rand_device, dtype=dtype) for i in range(batch_size)
```
Comment (anton-l) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:389]: Looks neat :)

**Review by anton-l — APPROVED** [review]
LGTM! Not sure about updating the types and docs for incompatible pipelines like DDPM and K-Diffusion though, but depends on how quickly we add the feature there afterwards :)

File: src/diffusers/pipelines/stable_diffusion/pipeline_cycle_diffusion.py [hunk 3]
```diff
@@ -435,8 +435,22 @@ def get_timesteps(self, num_inference_steps, strength, device):
...
     def prepare_latents(self, image, timestep, batch_size, num_images_per_prompt, dtype, device, generator=None):
         image = image.to(device=device, dtype=dtype)
-        init_latent_dist = self.vae.encode(image).latent_dist
-        init_latents = init_latent_dist.sample(generator=generator)
+
+        batch_size = image.shape[0]
+        if isinstance(generator, list) and len(generator) != batch_size:
+            raise ValueError(
+                f"You have passed a list of generators of length {len(generator)}, but requested an effective batch"
+                f" size of {batch_size}. Make sure the batch sizes matches the length of the generators."
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_cycle_diffusion.py:443]:
```suggestion
                f" size of {batch_size}. Make sure the batch size matches the length of the generators."
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py [hunk 4]
```diff
@@ -345,8 +345,22 @@ def get_timesteps(self, num_inference_steps, strength, device):
...
     def prepare_latents(self, image, timestep, batch_size, num_images_per_prompt, dtype, device, generator=None):
         image = image.to(device=device, dtype=dtype)
-        init_latent_dist = self.vae.encode(image).latent_dist
-        init_latents = init_latent_dist.sample(generator=generator)
+
+        batch_size = batch_size * num_images_per_prompt
+        if isinstance(generator, list) and len(generator) != batch_size:
+            raise ValueError(
+                f"You have passed a list of generators of length {len(generator)}, but requested an effective batch"
+                f" size of {batch_size}. Make sure the batch sizes matches the length of the generators."
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:353]:
```suggestion
                f" size of {batch_size}. Make sure the batch size matches the length of the generators."
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_image_variation.py [hunk 5]
```diff
@@ -258,12 +258,24 @@ def check_inputs(self, image, height, width, callback_steps):
     # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.prepare_latents
     def prepare_latents(self, batch_size, num_channels_latents, height, width, dtype, device, generator, latents=None):
         shape = (batch_size, num_channels_latents, height // self.vae_scale_factor, width // self.vae_scale_factor)
+        if isinstance(generator, list) and len(generator) != batch_size:
+            raise ValueError(
+                f"You have passed a list of generators of length {len(generator)}, but requested an effective batch"
+                f" size of {batch_size}. Make sure the batch sizes matches the length of the generators."
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_image_variation.py:264]:
```suggestion
                f" size of {batch_size}. Make sure the batch size matches the length of the generators."
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py [hunk 6]
```diff
@@ -401,8 +401,22 @@ def get_timesteps(self, num_inference_steps, strength, device):
...
     def prepare_latents(self, image, timestep, batch_size, num_images_per_prompt, dtype, device, generator=None):
         image = image.to(device=device, dtype=dtype)
-        init_latent_dist = self.vae.encode(image).latent_dist
-        init_latents = init_latent_dist.sample(generator=generator)
+
+        batch_size = batch_size * num_images_per_prompt
+        if isinstance(generator, list) and len(generator) != batch_size:
+            raise ValueError(
+                f"You have passed a list of generators of length {len(generator)}, but requested an effective batch"
+                f" size of {batch_size}. Make sure the batch sizes matches the length of the generators."
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py:409]:
```suggestion
                f" size of {batch_size}. Make sure the batch size matches the length of the generators."
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py [hunk 7]
```diff
@@ -463,12 +463,24 @@ def check_inputs(self, prompt, height, width, callback_steps):
     # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.prepare_latents
     def prepare_latents(self, batch_size, num_channels_latents, height, width, dtype, device, generator, latents=None):
         shape = (batch_size, num_channels_latents, height // self.vae_scale_factor, width // self.vae_scale_factor)
+        if isinstance(generator, list) and len(generator) != batch_size:
+            raise ValueError(
+                f"You have passed a list of generators of length {len(generator)}, but requested an effective batch"
+                f" size of {batch_size}. Make sure the batch sizes matches the length of the generators."
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_inpaint.py:469]:
```suggestion
                f" size of {batch_size}. Make sure the batch size matches the length of the generators."
```

File: src/diffusers/pipelines/stable_diffusion_safe/pipeline_stable_diffusion_safe.py [hunk 8]
```diff
@@ -422,12 +422,24 @@ def check_inputs(self, prompt, height, width, callback_steps):
     # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.prepare_latents
     def prepare_latents(self, batch_size, num_channels_latents, height, width, dtype, device, generator, latents=None):
         shape = (batch_size, num_channels_latents, height // self.vae_scale_factor, width // self.vae_scale_factor)
+        if isinstance(generator, list) and len(generator) != batch_size:
+            raise ValueError(
+                f"You have passed a list of generators of length {len(generator)}, but requested an effective batch"
+                f" size of {batch_size}. Make sure the batch sizes matches the length of the generators."
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion_safe/pipeline_stable_diffusion_safe.py:428]:
```suggestion
                f" size of {batch_size}. Make sure the batch size matches the length of the generators."
```

File: src/diffusers/pipelines/versatile_diffusion/pipeline_versatile_diffusion_dual_guided.py [hunk 9]
```diff
@@ -376,12 +376,24 @@ def check_inputs(self, prompt, image, height, width, callback_steps):
     # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.prepare_latents
     def prepare_latents(self, batch_size, num_channels_latents, height, width, dtype, device, generator, latents=None):
         shape = (batch_size, num_channels_latents, height // self.vae_scale_factor, width // self.vae_scale_factor)
+        if isinstance(generator, list) and len(generator) != batch_size:
+            raise ValueError(
+                f"You have passed a list of generators of length {len(generator)}, but requested an effective batch"
+                f" size of {batch_size}. Make sure the batch sizes matches the length of the generators."
```
Comment (pcuenca) [inline src/diffusers/pipelines/versatile_diffusion/pipeline_versatile_diffusion_dual_guided.py:382]:
```suggestion
                f" size of {batch_size}. Make sure the batch size matches the length of the generators."
```

File: src/diffusers/pipelines/versatile_diffusion/pipeline_versatile_diffusion_image_variation.py [hunk 10]
```diff
@@ -241,12 +241,24 @@ def check_inputs(self, image, height, width, callback_steps):
     # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.prepare_latents
     def prepare_latents(self, batch_size, num_channels_latents, height, width, dtype, device, generator, latents=None):
         shape = (batch_size, num_channels_latents, height // self.vae_scale_factor, width // self.vae_scale_factor)
+        if isinstance(generator, list) and len(generator) != batch_size:
+            raise ValueError(
+                f"You have passed a list of generators of length {len(generator)}, but requested an effective batch"
+                f" size of {batch_size}. Make sure the batch sizes matches the length of the generators."
```
Comment (pcuenca) [inline src/diffusers/pipelines/versatile_diffusion/pipeline_versatile_diffusion_image_variation.py:247]:
```suggestion
                f" size of {batch_size}. Make sure the batch size matches the length of the generators."
```

File: src/diffusers/pipelines/versatile_diffusion/pipeline_versatile_diffusion_text_to_image.py [hunk 11]
```diff
@@ -292,12 +292,24 @@ def check_inputs(self, prompt, height, width, callback_steps):
     # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline.prepare_latents
     def prepare_latents(self, batch_size, num_channels_latents, height, width, dtype, device, generator, latents=None):
         shape = (batch_size, num_channels_latents, height // self.vae_scale_factor, width // self.vae_scale_factor)
+        if isinstance(generator, list) and len(generator) != batch_size:
+            raise ValueError(
+                f"You have passed a list of generators of length {len(generator)}, but requested an effective batch"
+                f" size of {batch_size}. Make sure the batch sizes matches the length of the generators."
```
Comment (pcuenca) [inline src/diffusers/pipelines/versatile_diffusion/pipeline_versatile_diffusion_text_to_image.py:298]:
```suggestion
                f" size of {batch_size}. Make sure the batch size matches the length of the generators."
```

File: tests/test_pipelines_common.py [hunk 12]
```diff
@@ -176,6 +181,64 @@ def test_inference_batch_consistent(self):
...
+        output_batch = pipe(**batched_inputs)
+        assert output_batch[0].shape[0] == batch_size
+
+        inputs["generator"] = self.get_generator(0)
+
+        output = pipe(**inputs)
+
+        if torch_device != "mps":
+            # TODO(Pedro) - not sure why, but not at all reproducible at the moment it seems
+            # make sure that batched and non-batched is identical
```
Comment (pcuenca) [inline tests/test_pipelines_common.py:238]: This is possibly because of https://github.com/pytorch/pytorch/issues/84039#issuecomment-1281125012. We can't really support more than 1 image on mps yet.

**Review by pcuenca — APPROVED** [review]
This should be very useful, thanks a lot!

File: src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_flax_stable_diffusion.py]: Good catch - removed it

**Comment by xiankgx** [comment]
For a batched generation, how do I use the generator? I passed in a list of generators with the same manual seed number but the first image differs from the other images in a batch? If I put in just one generator, then all the images in the batch are different. But using the same seed results in the same batch of images.

**Comment by patrickvonplaten** [comment]
Hey @xiankgx,

Could you please open a new issue with a reproducible code snippet?
